// contentful-rich-text-vue-renderer/dist/index.esm.js
import { MARKS, BLOCKS, INLINES, helpers } from '@contentful/rich-text-types';

var defaultInline = function (type, node, key, h) {
    return h(
        'span',
        {
            key: key,
            style: {
                margin: "0px 5px",
                padding: "0 .25rem 0 .75rem",
                border: "1px solid #d3dce0",
                fontFamily: "monospace"
            },
        },
        ("inline: " + type + ", sys.id: " + (node.data.target.sys.id))
    );
};

var defaultMarkRenderers = {};
defaultMarkRenderers[MARKS.BOLD] = function (text, key, h) { return h('strong', { key: key }, text); };
defaultMarkRenderers[MARKS.ITALIC] = function (text, key, h) { return h('em', { key: key }, text); };
defaultMarkRenderers[MARKS.UNDERLINE] = function (text, key, h) { return h('u', { key: key }, text); };
defaultMarkRenderers[MARKS.CODE] = function (text, key, h) { return h('code', { key: key }, text); };

var defaultNodeRenderers = {};
defaultNodeRenderers[BLOCKS.PARAGRAPH] = function (node, key, h, next) { return (
        h('p', { key: key }, next(node.content, key, h, next))
    ); };
defaultNodeRenderers[BLOCKS.HEADING_1] = function (node, key, h, next) { return (
        h('h1', { key: key }, next(node.content, key, h, next))
    ); };
defaultNodeRenderers[BLOCKS.HEADING_2] = function (node, key, h, next) { return (
        h('h2', { key: key }, next(node.content, key, h, next))
    ); };
defaultNodeRenderers[BLOCKS.HEADING_3] = function (node, key, h, next) { return (
        h('h3', { key: key }, next(node.content, key, h, next))
    ); };
defaultNodeRenderers[BLOCKS.HEADING_4] = function (node, key, h, next) { return (
        h('h4', { key: key }, next(node.content, key, h, next))
    ); };
defaultNodeRenderers[BLOCKS.HEADING_5] = function (node, key, h, next) { return (
        h('h5', { key: key }, next(node.content, key, h, next))
    ); };
defaultNodeRenderers[BLOCKS.HEADING_6] = function (node, key, h, next) { return (
        h('h6', { key: key }, next(node.content, key, h, next))
    ); };
defaultNodeRenderers[BLOCKS.EMBEDDED_ENTRY] = function (node, key, h, next) { return (
        h('div', { key: key }, next(node.content, key, h, next))
    ); };
defaultNodeRenderers[BLOCKS.UL_LIST] = function (node, key, h, next) { return (
        h('ul', { key: key }, next(node.content, key, h, next))
    ); };
defaultNodeRenderers[BLOCKS.OL_LIST] = function (node, key, h, next) { return (
        h('ol', { key: key }, next(node.content, key, h, next))
    ); };
defaultNodeRenderers[BLOCKS.LIST_ITEM] = function (node, key, h, next) { return (
        h('li', { key: key }, next(node.content, key, h, next))
    ); };
defaultNodeRenderers[BLOCKS.QUOTE] = function (node, key, h, next) { return (
        h('blockquote', { key: key }, next(node.content, key, h, next))
    ); };
defaultNodeRenderers[BLOCKS.HR] = function (_node, key, h) { return h('hr', { key: key }, {}); };
defaultNodeRenderers[INLINES.ASSET_HYPERLINK] = function (node, key, h) { return defaultInline(INLINES.ASSET_HYPERLINK, node, key, h); };
defaultNodeRenderers[INLINES.ENTRY_HYPERLINK] = function (node, key, h) { return defaultInline(INLINES.ENTRY_HYPERLINK, node, key, h); };
defaultNodeRenderers[INLINES.EMBEDDED_ENTRY] = function (node, key, h) { return defaultInline(INLINES.EMBEDDED_ENTRY, node, key, h); };
defaultNodeRenderers[INLINES.HYPERLINK] = function (node, key, h, next) {
        return h(
            'a',
            {
                key: key,
                attrs: {
                  href: node.data.uri
                }
            },
            next(node.content, key, h, next)
        )
    };
defaultNodeRenderers.text = function (ref, key, h, markRenderer) {
        var marks = ref.marks;
        var value = ref.value;

        return marks.length
            ? marks.reduce(
                  function (aggregate, mark, i) { return markRenderer[mark.type](aggregate, (key + "-" + i), h); },
                  value
              )
            : value;
    };

var renderNodeList = function (nodes, key, renderer) {
    return nodes.map(function (node, i) { return renderNode(node, (key + "-" + i), renderer); });
};

var renderNode = function (node, key, renderer) {
    var nodeRenderer = renderer.node;
    var createElement = renderer.createElement;

    if (helpers.isText(node)) {
        // We're at final tip of node branch, can render text.
        var markerRender = renderer.mark;
        return nodeRenderer.text(node, key, createElement, markerRender);
    } else {
        var nextNode = function (nodes) { return renderNodeList(nodes, key, renderer); };
        if (!nodeRenderer) {
            return createElement('div', (key + " ;lost nodeRenderer"));
        }
        if (!node.nodeType || !nodeRenderer[node.nodeType]) {
            // TODO: Figure what to return when passed an unrecognized node.
            return "(Unrecognized node type) " + (node.nodeType || "empty");
        }
        return nodeRenderer[node.nodeType](node, key, createElement, nextNode);
    }
};

var index = {
    functional: true,

    props: ["document", "nodeRenderers", "markRenderers"],

    render: function render(h, ctx) {
        var renderer = {
            node: Object.assign({}, defaultNodeRenderers,
                ctx.props.nodeRenderers),
            mark: Object.assign({}, defaultMarkRenderers,
                ctx.props.markRenderers),
            createElement: h
        };

        return renderNodeList(ctx.props.document.content, "RichText-", renderer);
    }
};

export default index;
