// contentful-rich-text-vue-renderer/dist/index.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@contentful/rich-text-types')) :
    typeof define === 'function' && define.amd ? define(['exports', '@contentful/rich-text-types'], factory) :
    (global = global || self, factory(global.ContentfulRichText = {}, global.richTextTypes));
}(this, function (exports, richTextTypes) { 'use strict';

    var defaultInline = function (type, node, key, h) {
        return h(
            'span',
            {
                key: key,
                style: {
                    margin: "0px 5px",
                    padding: "0 .25rem 0 .75rem",
                    border: "1px solid #d3dce0",
                    fontFamily: "monospace"
                },
            },
            ("inline: " + type + ", sys.id: " + (node.data.target.sys.id))
        );
    };

    var defaultMarkRenderers = {};
    defaultMarkRenderers[richTextTypes.MARKS.BOLD] = function (text, key, h) { return h('strong', { key: key }, text); };
    defaultMarkRenderers[richTextTypes.MARKS.ITALIC] = function (text, key, h) { return h('em', { key: key }, text); };
    defaultMarkRenderers[richTextTypes.MARKS.UNDERLINE] = function (text, key, h) { return h('u', { key: key }, text); };
    defaultMarkRenderers[richTextTypes.MARKS.CODE] = function (text, key, h) { return h('code', { key: key }, text); };

    var defaultNodeRenderers = {};
    defaultNodeRenderers[richTextTypes.BLOCKS.PARAGRAPH] = function (node, key, h, next) { return (
            h('p', { key: key }, next(node.content, key, h, next))
        ); };
    defaultNodeRenderers[richTextTypes.BLOCKS.HEADING_1] = function (node, key, h, next) { return (
            h('h1', { key: key }, next(node.content, key, h, next))
        ); };
    defaultNodeRenderers[richTextTypes.BLOCKS.HEADING_2] = function (node, key, h, next) { return (
            h('h2', { key: key }, next(node.content, key, h, next))
        ); };
    defaultNodeRenderers[richTextTypes.BLOCKS.HEADING_3] = function (node, key, h, next) { return (
            h('h3', { key: key }, next(node.content, key, h, next))
        ); };
    defaultNodeRenderers[richTextTypes.BLOCKS.HEADING_4] = function (node, key, h, next) { return (
            h('h4', { key: key }, next(node.content, key, h, next))
        ); };
    defaultNodeRenderers[richTextTypes.BLOCKS.HEADING_5] = function (node, key, h, next) { return (
            h('h5', { key: key }, next(node.content, key, h, next))
        ); };
    defaultNodeRenderers[richTextTypes.BLOCKS.HEADING_6] = function (node, key, h, next) { return (
            h('h6', { key: key }, next(node.content, key, h, next))
        ); };
    defaultNodeRenderers[richTextTypes.BLOCKS.EMBEDDED_ENTRY] = function (node, key, h, next) { return (
            h('div', { key: key }, next(node.content, key, h, next))
        ); };
    defaultNodeRenderers[richTextTypes.BLOCKS.UL_LIST] = function (node, key, h, next) { return (
            h('ul', { key: key }, next(node.content, key, h, next))
        ); };
    defaultNodeRenderers[richTextTypes.BLOCKS.OL_LIST] = function (node, key, h, next) { return (
            h('ol', { key: key }, next(node.content, key, h, next))
        ); };
    defaultNodeRenderers[richTextTypes.BLOCKS.LIST_ITEM] = function (node, key, h, next) { return (
            h('li', { key: key }, next(node.content, key, h, next))
        ); };
    defaultNodeRenderers[richTextTypes.BLOCKS.QUOTE] = function (node, key, h, next) { return (
            h('blockquote', { key: key }, next(node.content, key, h, next))
        ); };
    defaultNodeRenderers[richTextTypes.BLOCKS.HR] = function (_node, key, h) { return h('hr', { key: key }, {}); };
    defaultNodeRenderers[richTextTypes.INLINES.ASSET_HYPERLINK] = function (node, key, h) { return defaultInline(richTextTypes.INLINES.ASSET_HYPERLINK, node, key, h); };
    defaultNodeRenderers[richTextTypes.INLINES.ENTRY_HYPERLINK] = function (node, key, h) { return defaultInline(richTextTypes.INLINES.ENTRY_HYPERLINK, node, key, h); };
    defaultNodeRenderers[richTextTypes.INLINES.EMBEDDED_ENTRY] = function (node, key, h) { return defaultInline(richTextTypes.INLINES.EMBEDDED_ENTRY, node, key, h); };
    defaultNodeRenderers[richTextTypes.INLINES.HYPERLINK] = function (node, key, h, next) {
            return h(
                'a',
                {
                    key: key,
                    attrs: {
                      href: node.data.uri
                    }
                },
                next(node.content, key, h, next)
            )
        };
    defaultNodeRenderers.text = function (ref, key, h, markRenderer) {
            var marks = ref.marks;
            var value = ref.value;

            return marks.length
                ? marks.reduce(
                      function (aggregate, mark, i) { return markRenderer[mark.type](aggregate, (key + "-" + i), h); },
                      value
                  )
                : value;
        };

    var renderNodeList = function (nodes, key, renderer) {
        return nodes.map(function (node, i) { return renderNode(node, (key + "-" + i), renderer); });
    };

    var renderNode = function (node, key, renderer) {
        var nodeRenderer = renderer.node;
        var createElement = renderer.createElement;

        if (richTextTypes.helpers.isText(node)) {
            // We're at final tip of node branch, can render text.
            var markerRender = renderer.mark;
            return nodeRenderer.text(node, key, createElement, markerRender);
        } else {
            var nextNode = function (nodes) { return renderNodeList(nodes, key, renderer); };
            if (!nodeRenderer) {
                return createElement('div', (key + " ;lost nodeRenderer"));
            }
            if (!node.nodeType || !nodeRenderer[node.nodeType]) {
                // TODO: Figure what to return when passed an unrecognized node.
                return "(Unrecognized node type) " + (node.nodeType || "empty");
            }
            return nodeRenderer[node.nodeType](node, key, createElement, nextNode);
        }
    };

    var index = {
        functional: true,

        props: ["document", "nodeRenderers", "markRenderers"],

        render: function render(h, ctx) {
            var renderer = {
                node: Object.assign({}, defaultNodeRenderers,
                    ctx.props.nodeRenderers),
                mark: Object.assign({}, defaultMarkRenderers,
                    ctx.props.markRenderers),
                createElement: h
            };

            return renderNodeList(ctx.props.document.content, "RichText-", renderer);
        }
    };

    exports.default = index;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
